### PR TITLE
Enhancement: implement document link custom field

### PR DESCRIPTION
**Pull request by shamoon**
## Proposed change

Screenshot 2023-12-03 at 11 03 13 PM
Screenshot 2023-12-04 at 7 59 31 PM

This was a fun little thing to do, most of it is frontend and made easier by one of our lovelier dependencies, `ng-select`, mostly seemed like hooking existing things up. Some notes:

- ~~Biggest debate I had was what type of Django field to use, I didnt want to deal with ManyToMany drama etc. so ultimately I decided on a list of the doc IDs, closest way to store that is a char field with the validator, I then parse it with JSON to make the API output cleaner.~~ Edit: JSONField. Also I chose to re-create the migration 1042 since its unreleased
- When loading the value we have to hit the API for the documents (e.g. title) async, as well as searching
- Can search by title, I thought search by ID (like GH) while cool, seemed unnecessary

Welcome any feedback.

Fixes #422

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Other (please explain): Enhancement

## Checklist:

- [x] I have read & agree with the [contributing guidelines](https://github.com/paperless-ngx/paperless-ngx/blob/main/CONTRIBUTING.md).
- [x] If applicable, I have included testing coverage for new code in this PR, for [backend](https://docs.paperless-ngx.com/development/#testing) and / or [front-end](https://docs.paperless-ngx.com/development/#testing-and-code-style) changes.
- [x] If applicable, I have tested my code for new features & regressions on both mobile & desktop devices, using the latest version of major browsers.
- [x] If applicable, I have checked that all tests pass, see [documentation](https://docs.paperless-ngx.com/development/#back-end-development).
- [x] I have run all `pre-commit` hooks, see [documentation](https://docs.paperless-ngx.com/development/#code-formatting-with-pre-commit-hooks).
- [x] I have made corresponding changes to the documentation as needed.
- [x] I have checked my modifications for any breaking changes.

### CODE REVIEW

File: src/documents/serialisers.py [hunk 1]
```diff
@@ -458,7 +460,11 @@ def create(self, validated_data):
         return instance
 
     def get_value(self, obj: CustomFieldInstance):
-        return obj.value
+        return (
+            obj.value
+            if (obj.field.data_type != CustomField.FieldDataType.DOCUMENTLINK)
+            else json.loads(obj.value)
```
Comment (stumpylog) [inline src/documents/serialisers.py:466]: Since it looks like you are using the new field as a JSON field, why not use a `JSONField`?

**Comment by shamoon** [comment]
> Any chance to add the correspondent and document creation date to the dropdown as well?

I think we need to be very sparing about what data we show in the dropdown as it’s not big. I’ll look at showing the date but correspondent seems arbitrary, other people will want doc type, etc.

**Comment by shamoon** [comment]
Ok, I converted to JSON field, also added the created date to the filter list, which I think is useful

LMK if anything else!

Screenshot 2023-12-04 at 7 59 31 PM

**Comment by jord1e** [comment]
🙏 maybe call it “References”?

**Comment by shamoon** [comment]
Its a custom field, you can give it any name you want.

**Review by stumpylog — APPROVED** [review]

**Comment by stumpylog** [comment]
Codecov seems confused here, feel free to ignore

**Comment by shamoon** [comment]
Ok yea I just noticed that, seems it’s just being weird 👍

**Comment by jord1e** [comment]
Create a custom field in the left menu and then in the top right on the document page click on "custom fields" to add it

grafik

**Pull request merged**